### ISSUE TITLE
Distinguish between primitive and object types in generic type constraints?

### ISSUE THREAD

**Issue by sccolbert** [Microsoft/TypeScript#3315]
With the introduction of WeakMap, the JS runtime now has [restrictions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap#Description) on where primitive types are disallowed. Is it possible (or planned) to have language support for discriminating from primitive types and the ability to constraint generic types based on that discrimination? 

From a [previous SO post I made](http://stackoverflow.com/questions/28122754/discriminate-between-empty-object-type-and-other-concrete-types), the answer seems to be "no". But taken in light of WeakMap requirements, perhaps it can be revisited?

```javascript
var a = 1;
undefined

var b = new Number(a);
undefined

var m = new WeakMap();
undefined

m.set(a, 1)
VM674:2 Uncaught TypeError: Invalid value used as weak map key
    at WeakMap.set (native)
    at <anonymous>:2:3
    at Object.InjectedScript._evaluateOn (<anonymous>:905:140)
    at Object.InjectedScript._evaluateAndWrap (<anonymous>:838:34)
    at Object.InjectedScript.evaluate (<anonymous>:694:21)(anonymous function) @ VM674:2InjectedScript._evaluateOn @ VM305:905InjectedScript._evaluateAndWrap @ VM305:838InjectedScript.evaluate @ VM305:694

m.set(b, 1)
WeakMap {Number {[[PrimitiveValue]]: 1} => 1}
```

**Comment by kitsonk** [comment]
It seems like you are asking for #1809 right?

**Comment by sccolbert** [comment]
@kitsonk yes, I think #1809 would handle it.

**Comment by jasonwilliams200OK** [comment]
> yes, I think #1809 would handle it.

In which case, I think this should be closed as duplicate of #1809. What do you think about that?

**Comment by sccolbert** [comment]
Agreed. This is also a duplicate of #2057.
